### PR TITLE
Misc meson fixes

### PR DESCRIPTION
**Pull request by evelikov**
Nothing overly fancy here, small cleanup proper docs/doxygen handling and a fancy summary()

@ceyusa as the original author of the meson build, do you mind having a look?

### CODE REVIEW

File: meson.build [hunk 1]
```diff
@@ -135,8 +135,8 @@ endif
 subdir('va')
 subdir('pkgconfig')
 
-doxygen = find_program('doxygen', required: false)
+doxygen = find_program('doxygen', required : get_option('enable_docs') == true)
```
Comment (eli-schwartz) [inline meson.build:151]: If enable_docs is true, then this evaluates to `true == true`, which in turn evaluates to true

If enable_docs is false, then this evaluates to `false == true`, which in turn evaluates to false.

You can just use the option directly. :)

File: meson.build [hunk 2]
```diff
@@ -140,3 +153,13 @@ doxygen = find_program('doxygen', required : get_option('enable_docs') == true)
 if get_option('enable_docs')
   subdir('doc')
 endif
+
+if meson.version().version_compare('>= 0.53')
```
Comment (eli-schwartz) [inline meson.build:157]: The minimum meson version is already this.

**Review by eli-schwartz — CHANGES_REQUESTED** [review]
Honestly, I'd use feature options more here.

File: meson.build (hunk 2 above)
Comment (evelikov) [reply meson.build]: Thanks for the help. Updated PR should have them addressed.

**Comment by eli-schwartz** [comment]
There is also https://mesonbuild.com/Build-options.html#deprecated-options but then you'd have to bump the minimum meson to something very new. :(

**Comment by eli-schwartz** [comment]
What would be the advantage of the synonym, other than the infrequent case that one migrates from a custom combo option that uses those strings?

File: va/meson.build [hunk 3]
```diff
@@ -101,12 +101,12 @@ if WITH_DRM
...
 
   libva_drm_args = []
-  if get_option('with_x11') != 'no' and x11_dep.found()
+  if WITH_X11
     libva_drm_sources += [ 'drm/va_drm_auth_x11.c' ]
     libva_drm_args += [
       '-DLIBVA_MAJOR_VERSION=@0@'.format(libva_major_version)
     ]
-    deps += [ x11_dep ]
+    deps += [ x11_deps ]
```
Comment (ceyusa) [inline va/meson.build:109]: I'm not sure if makes sens to link with all x11-related dependencies for drm

File: meson.build [hunk 4]
```diff
@@ -127,8 +138,16 @@ endif
 subdir('va')
 subdir('pkgconfig')
 
-doxygen = find_program('doxygen', required: false)
+doxygen = find_program('doxygen', required : get_option('enable_docs'))
 
-if get_option('enable_docs') and doxygen.found()
+if get_option('enable_docs')
```
Comment (ceyusa) [inline meson.build:143]: what if enable_docs is auto but doxygen is not installed?

**Comment by ceyusa** [comment]
It's a nice clean up. An initial step to modernize the meson build.

File: meson.build [hunk 5]
```diff
@@ -75,60 +75,50 @@ configinc = include_directories('.')
...
 
+# The libdrm package is required by all backends
 libdrm_dep = dependency('libdrm', version : '>= 2.4.60')
 
 backends = ''
-WITH_DRM = not get_option('disable_drm')
-if WITH_DRM
-  backends += 'drm'
+WITH_DRM = false
+if libdrm_dep.found() and not get_option('with_drm').disabled()
```
Comment (eli-schwartz) [inline meson.build:83]: libdrm_dep cannot ever be not-found, so this one is a false choice

File: meson.build (hunk 4 above)
Comment (eli-schwartz) [reply meson.build:143]: This is currently broken and should be switched to use `if doxygen.found()`.

**Review by eli-schwartz — CHANGES_REQUESTED** [review]

File: va/meson.build (hunk 3 above)
Comment (evelikov) [reply va/meson.build:109]: By default meson will pass `-Wl,--as-needed`, thus the linker will discard any unneeded DSOs.
Plus it makes little sense to have a separate dependency, just for this.

File: meson.build (hunk 4 above)
Comment (evelikov) [reply meson.build:143]: The enable_docs can never be auto - it's a bool.

Comment (evelikov) [reply meson.build:143]: > This is currently broken and should be switched to use `if doxygen.found()`.

```
$ meson -D enable_docs=true

Program doxygen found: NO

meson.build:141:0: ERROR: Program 'doxygen' not found or not executable
```

Seems to be working as expected. Note the option is a bool and not a tristate (combo or feature).

File: meson.build (hunk 5 above)
Comment (evelikov) [reply meson.build:83]: Correct libdrm_dep.found is always true. I've added it for symmetry with the other code-paths and it's not causing any issues - be that confusing meson or adding any measurable overhead.

**Review by dvrogozh — CHANGES_REQUESTED** [review]

File: va/meson.build (hunk 3 above)
Comment (evelikov) [reply va/meson.build:109]: I'm actually more tempted at removing the xlib dependency in the first place.

File: meson.build [hunk 6]
```diff
@@ -81,11 +81,19 @@ WITH_DRM = not get_option('disable_drm')
 
 WITH_X11 = false
 if get_option('with_x11') != 'no'
-  x11_dep = dependency('x11', required : get_option('with_x11') == 'yes')
-  xext_dep = dependency('xext', required : get_option('with_x11') == 'yes')
-  xfixes_dep = dependency('xfixes', required : get_option('with_x11') == 'yes')
-
-  WITH_X11 = (x11_dep.found() and xext_dep.found() and xfixes_dep.found())
+  require_x11 =  get_option('with_x11') == 'yes'
```
[the remark this answers is not in the document]
Comment (evelikov) [reply meson.build]: Having a simple foreach loop is idiomatic approach when using meson. Having it open-coded is repetitive and error-prone. Things will be even neater with the DRI3/Xwayland MR.

File: va/meson.build (hunk 3 above)
Comment (evelikov) [reply va/meson.build]: Dependency removed. This hunk is no longer relevant.

File: meson.build (hunk 6 above)
[the remark this answers is not in the document]
Comment (evelikov) [reply meson.build]: If you're interested at shorter code, then you'll love the last commit - using meson features. If you haven't used them before the [documentation](https://mesonbuild.com/Build-options.html#features) describes them well better than I could.

**Review by XinfengZhang — APPROVED** [review]

File: va/drm/va_drm_auth_x11.c [hunk 7]
```diff
@@ -25,9 +25,16 @@
 #define _GNU_SOURCE 1
 #include "sysdeps.h"
 #include <dlfcn.h>
-#include <X11/Xlib.h>
+#include <stdint.h>
 #include "va_drm_auth_x11.h"
 
+/* A few ABI compabible defines and typedefs to avoid pulling Xlib.h */
```
Comment (dvrogozh) [inline va/drm/va_drm_auth_x11.c:31]: If we need Xlib.h, then we need to include it rather than mimic it. If we don't need to link against the library, then don't link, but add paths to include files - this should be managed on build system level.

**Review by dvrogozh — CHANGES_REQUESTED** [review]

File: va/drm/va_drm_auth_x11.c [hunk 8]
```diff
@@ -115,7 +122,7 @@ drm_auth_x11_init(DRMAuthX11 *auth)
     if (!auth->display)
         return false;
 
-    auth->window = DefaultRootWindow(auth->display);
+    auth->window = None;
```
Comment (dvrogozh) [inline va/drm/va_drm_auth_x11.c:125]: Why this is equivalent to what was here before?

File: va/drm/va_drm_auth_x11.c (hunk 7 above)
Comment (evelikov) [reply va/drm/va_drm_auth_x11.c:31]: I would greatly encourage you to take a dive in the the Xlib.h headers. They are one of the worst one that exist. In practice this is the least crazy thing to do here.

File: va/drm/va_drm_auth_x11.c (hunk 8 above)
Comment (evelikov) [reply va/drm/va_drm_auth_x11.c:125]: It is unused, so for all intents and purposes we can make it anything that we want.

**Comment by evelikov** [comment]
@XinfengZhang any objections if we merge this for the next release? Obviously I would need to rebase it on top of the va-win32 addition that just landed.

**Comment by evelikov** [comment]
Rebased the the PR with couple of small changes:
 - pushed the "use meson features" earlier in the series
 - make win32 a feature like the rest cc @sivileri

@XinfengZhang @dvrogozh as mentioned earlier, I would love it this can make it for 2.17.

**Comment by evelikov** [comment]
@XinfengZhang @dvrogozh humble ping? Is there anything I can do to move this forward?

**Comment by evelikov** [comment]
Bear in mind that the last commit will conflict, albeit trivially, with https://github.com/intel/libva/pull/656

**Comment by evelikov** [comment]
@XinfengZhang @dvrogozh humble ping?

**Review by dvrogozh — CHANGES_REQUESTED** [review]

File: va/drm/va_drm_auth_x11.c (hunk 7 above)
Comment (evelikov) [reply va/drm/va_drm_auth_x11.c:31]: Sounds like throwing adjectives, instead of describing the details was a little misleading. Let me try again:

Various parts of X, predate standard C89/C99 and as such they often had to invent them to a varying degree. In some cases the X implementations would ignore/violate C standards be compatible on the API level but vary in ABI. Furthermore often the whole internal X ABI would be leaked and thus embedded into unsuspecting programs.

In this case here, we need a few very well defined and stable (from CPU arch or *nix flavour POV) types. We simply do not need the chaos that `#include <X11/Xlib.h>` brings.

For the same reasons, we've done similar changes in Mesa, so I believe we could and should really do the same in here. Does that make sense?

Comment (evelikov) [reply va/drm/va_drm_auth_x11.c:31]: @dvrogozh ping?

File: meson.build [hunk 9]
```diff
@@ -75,44 +75,62 @@ configinc = include_directories('.')
 cc = meson.get_compiler('c')
 dl_dep = cc.find_library('dl', required : false)
 
-WITH_DRM = not get_option('disable_drm') and (host_machine.system() != 'windows')
-libdrm_dep = dependency('libdrm', version : '>= 2.4.60', required : (host_machine.system() != 'windows'))
+backends = ''
+
+WITH_DRM = false
+if not get_option('with_drm').disabled()
+  libdrm_dep = dependency('libdrm', version : '>= 2.4.60', required : (host_machine.system() != 'windows'))
```
Comment (xclaesse) [inline meson.build:82]: If with_drm is set to auto, the libdrm dependency should be optional. Here you make it required on non-windows.

I would suggest doing instead:
```
libdrm_dep = dependency('libdrm', version : '>= 2.4.60', required : get_option('with_drm'))
WITH_DRM = libdrm_dep.found()
if WITH_DRM
  backends += ' drm'
endif
```

I don't the it matters to check the platform, if someone build with -Dwith_drm=enabled on Windows, they'll get an error because libdrm is not found and that's fine IMHO.

File: meson.build [hunk 10]
```diff
@@ -75,44 +75,62 @@ configinc = include_directories('.')
...
+  x11_deps = [
+    dependency('x11', required : get_option('with_x11')),
+    dependency('xext', required : get_option('with_x11')),
+    dependency('xfixes', required : get_option('with_x11')),
+    dependency('x11-xcb', required : get_option('with_x11')),
+    dependency('xcb', required : get_option('with_x11')),
+    dependency('xcb-dri3', required : get_option('with_x11')),
+  ]
+  foreach d : x11_deps
+    WITH_X11 = WITH_X11 and d.found()
```
Comment (xclaesse) [inline meson.build:100]: WITH_X11 is initially set to false, I think this will always be false then.

**Comment by xclaesse** [comment]
Other than those small comments, the meson part looks good to me, it's a nice cleanup.

File: meson.build (hunk 9 above)
Comment (evelikov) [reply meson.build:82]: > If with_drm is set to auto, the libdrm dependency should be optional. 

Not quite: libdrm is a hard requirement for all non-win32 backends - android, x11 (dri3) and by extension glx, wayland (wl_drm) or drm itself - they all use the `va_drm_utils.c` helper which needs libdrm.

>I don't the it matters to check the platform, if someone build with -Dwith_drm=enabled on Windows, they'll get an error because libdrm is not found and that's fine IMHO.

There is a word missing above, it seems. Regardless - as long as the author @sivileri is fine it works for me.

File: meson.build (hunk 10 above)
Comment (evelikov) [reply meson.build:100]: Thanks fixed :+1:

**Comment by evelikov** [comment]
@dvrogozh @XinfengZhang humble monthly ping?

File: meson.build (hunk 9 above)
Comment (sivileri) [reply meson.build:82]: For Windows it should still be fine, as long as nobody does `-Dwith_drm` without `libdrm`. I don't have any objections with making `libdrm` optional, as long as we are aware, as @evelikov points out, that doing the proposed change will make `libdrm` optional in all non-windows environments with have a hard dependency to it.

Comment (evelikov) [reply meson.build:82]: Sounds like we're fine as-is them.

Comment (xclaesse) [reply meson.build:84]: @evelikov I'm not libva expert, but it still feels wrong. With this PR, on linux, you can do `-Dwith_drm=disabled` and it will not use libdrm at all, so that dep is optional even on non-Windows platforms? If it makes sense to have libdrm disabled for a linux build, then `with_drm=auto` should allow libdrm dependency to be not-found. Only `-Dwith_drm=enabled` should force having libdrm on Linux IMHO. Otherwise a tristate enabled/disabled/auto does not make much sense, in your current implementation auto has the same effect as enabled. Unless there is something I missed?

Comment (xclaesse) [reply meson.build:84]: Possibly with_drm should not be an option at all and we can remove the ` if not get_option('with_drm').disabled()` condition, to not allow disabling libdrm?

Comment (evelikov) [reply meson.build:84]: Your wording threw me off there for a moment but you're spot on.

In particular `-D with_drm=disabled` is borked - the backend is correctly not selected for building, yet building the x11, wayland... ones fail, since meson sees the libdrm dependency as unresolved.

Will update in a moment to drop the explicit drm toggle - it makes zero sense to have it. Thanks for the suggestion and patience.

File: meson.build [hunk 11]
```diff
@@ -75,50 +75,70 @@ configinc = include_directories('.')
...
+# NOTE: libdrm is required by all non-win32 backends x11 via dri3, glx via x11,
+# wayland via wl_drm and drm. All of these use va_drm_util.c which need libdrm.
+libdrm_dep = dependency('libdrm', version : '>= 2.4.60', required : (host_machine.system() != 'windows'))
 
-  x11_xcb_dep = dependency('x11-xcb', required : get_option('with_x11') == 'yes')
-  xcb_dep = dependency('xcb', required : get_option('with_x11') == 'yes')
-  xcb_dri3_dep = dependency('xcb-dri3', required : get_option('with_x11') == 'yes')
-  WITH_X11 = (WITH_X11 and x11_xcb_dep.found() and xcb_dep.found() and xcb_dri3_dep.found())
+WITH_X11 = true
+if not get_option('with_x11').disabled()
```
Comment (xclaesse) [inline meson.build:89]: You have the opposite problem now, WITH_X11 is going to be true if with_x11=disabled AFAIK.

Comment (evelikov) [reply meson.build:89]: thanks - third time's the charm :crossed_fingers:

File: va/drm/va_drm_auth_x11.c (hunk 7 above)
Comment (evelikov) [reply va/drm/va_drm_auth_x11.c]: @XinfengZhang @dvrogozh humble poke
